### ISSUE TITLE
Calling `Promise.resolve()` inside of an iframe crashes the process

### ISSUE THREAD

**Issue by jlongster** [electron/electron#12467]
* Electron version: 1.8.4
* Operating system: macOS 10.13.3

### Expected behavior

The app shouldn't crash

### Actual behavior

The app crashes

### How to reproduce

In my app I'm using react-load-script to dynamically pull in external JS. It used to work, and I'm not sure when it stopped working, but I'm now facing a crash. If I comment out this line in the library, the app doesn't crash: https://github.com/blueberryapps/react-load-script/blob/master/src/index.jsx#L122

That line adds a dynamically created `<script src="..." async>` tag to the DOM. I'm using one `<Script>` tag at  a certain point in my app, and it crashes. Here's the crash details from my error tracking service:

<img width="839" alt="screen shot 2018-03-28 at 10 16 37 pm" src="https://user-images.githubusercontent.com/17031/38067718-c9514336-32db-11e8-9020-b97d1b8f10a0.png">

This is probably pretty obscure, but maybe somebody notices something that can help. I can also try to create a small reproducible test case and see if it crashes using this lib in fresh project.

**Comment by jlongster** [comment]
fwiw I was able to reproduce this on 2.0, unfortunately. I'm not blocked on this bug right now, but when I do need this fixed I will investigate it more deeply myself and report back. But if anyone knows where I should start looking, I'm all ears.

**Comment by nornagon** [comment]
Thanks for the report! Do you have a small test case that we could try reproducing with locally?

**Comment by jlongster** [comment]
Not yet, still working on it. It's hard to pin down exactly what the conditions are. I'll let you know.

**Comment by derek-duncan** [comment]
I am also experiencing the renderer crash when a script is appended to the DOM. This was working in 1.7.9, but upgrading to 2.0 is causing a crash. In my case, I'm loading in the Intercom script with their standard script.

```javascript
//Set your APP_ID
var APP_ID = "APP_ID";

(function(){var w=window;var ic=w.Intercom;if(typeof ic==="function"){ic('reattach_activator');ic('update',intercomSettings);}else{var d=document;var i=function(){i.c(arguments)};i.q=[];i.c=function(args){i.q.push(args)};w.Intercom=i;function l(){var s=d.createElement('script');s.type='text/javascript';s.async=true;
s.src='https://widget.intercom.io/widget/' + APP_ID;
var x=d.getElementsByTagName('script')[0];x.parentNode.insertBefore(s,x);}if(w.attachEvent){w.attachEvent('onload',l);}else{w.addEventListener('load',l,false);}}})()
```

**Comment by jlongster** [comment]
@nornagon I found a small repro. Evidently it doesn't have anything to do with loading the script tag. The problem is actually with the code that runs when the script is loaded.

I'm using [Plaid](https://plaid.com/), and am loading their client script. It crashes on the `Plaid.create` call. Unfortunately that means it's impossibly for me to debug any further. But the following code should crash electron as of 2.0.4:

```js
      const script = document.createElement('script');
      script.id = 'plaid-script';
      script.src = 'https://cdn.plaid.com/link/v2/stable/link-initialize.js';
      script.async = 1;

      script.onload = () => {
        window.Plaid.create({
          env: 'development',
          clientName: 'client',
          key: 'foo',
          product: ['transactions'],
          onSuccess: async (publicToken, metadata) => {
          },
          onExit: (err, metadata) => {
          }
        });
      };

      script.onerror = () => {
        console.log('errored');
      };

      document.body.appendChild(script);
```

Hopefully if this crashes for you, that gives you something to track down?

**Comment by jlongster** [comment]
Note that *was* also crashing for me on `1.8.4`. So it's not a 2.0-specific issue (in my case).

I'm on macOS.

Not sure if we want to rename the issue.

**Comment by jlongster** [comment]
I just tried it in Chrome 61 and it definitely doesn't work. It's something with the electron-specific Promise hooks or something. Still digging.

**Comment by jlongster** [comment]
Is there a way to attach LLDB to electron (specifically the render process)?

**Comment by jlongster** [comment]
I figured it out!

Here's a tiny reproducible test case. You need to use an iframe so you need two files:

index.html:
```html
<iframe src="/crash.html"></iframe>
```

crash.html:
```html
<script>Promise.resolve()</script>
```

For some reason, calling `Promise.resolve()` inside of an iframe crashes electron (both 1.x and the latest 2.x version). I'm updating the title of the issue to reflect this.

How hard do you think this is to fix?

**Comment by jlongster** [comment]
This works on version `1.7.15`. Something broke between `1.7.15` and `1.8.1` and has been broken since. Will see if there's any easy way to get related commits.

**Comment by jlongster** [comment]
Ooooh it looks like this is fixed in version 3.0.0-beta.1, which is great! Just means I'll have to stay on 1.7.15 until 3.0.0 is ready, but at least we know a fix is coming.

**Comment by ckerr** [comment]
Closing this ticket as the issue is fixed in `3.0.0-beta.1`
